src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,191 @@
+
+#include "Globals.h"
+
+#include "EnderDragonFightStructuresGen.h"
+#include "../Chunk.h"
+#include "../Entities/EnderCrystal.h"
+#include "../Mobs/EnderDragon.h"
+#include "../WorldStorage/SchematicFileSerializer.h"
+
+cEnderDragonFightStructuresGen::cEnderDragonFightStructuresGen(int a_Seed, const AString & a_TowerProperties, int a_Radius) :
+	m_Noise(a_Seed)
+{
+	// Loads the fountain schematic
+	if (!cSchematicFileSerializer::LoadFromSchematicFile(m_Fountain, AString("Prefabs") + cFile::GetPathSeparator() + "SinglePieceStructures" + cFile::GetPathSeparator() + "EndFountain.schematic"))
+	{
+		LOGWARNING("EnderDragonFightStructuresGen is missing it's end fountain prefab, please update your cuberite server files! There will be no end fountain!");
+	}
+	// Reads the given tower properties
+	const auto TowerPropertiesVector = StringSplitAndTrim(a_TowerProperties, ";");
+	for (const auto & TowerProperty : TowerPropertiesVector)
+	{
+		const auto TowerPropertyVector = StringSplitAndTrim(TowerProperty, "|");
+		if (TowerPropertyVector.size() != 3)
+		{
+			LOGWARNING("Got unmatching parameters on generating obsidian pillars: %s, Please use \"Height|Radius|HasCage\", ...", TowerProperty);
+			continue;
+		}
+		int Height = std::stoi(TowerPropertyVector[0]);
+		int Radius = std::stoi(TowerPropertyVector[1]);
+		bool HasCage;
+		if (NoCaseCompare(TowerPropertyVector[2], "true") == 0)
+		{
+			HasCage = true;
+		}
+		else if (NoCaseCompare(TowerPropertyVector[2], "false") == 0)
+		{
+			HasCage = false;
+		}
+		else
+		{
+			LOGWARNING("Got unknown value for boolean if the tower: %s should have a cage! %s", TowerProperty, TowerPropertyVector[2]);
+			continue;
+		}
+		m_TowerProperties.push_back({Height, Radius, HasCage});
+	}
+	// A random angle in radian
+	double Angle = m_Noise.IntNoise1D(a_Seed) * M_PI + M_PI;
+	// Generate Positions in a circle
+	for (int i = 0; i < static_cast<int>(m_TowerProperties.size()); i++)
+	{
+		auto TowerPos = Vector3i(FloorC(a_Radius * cos(Angle)), 0, FloorC(a_Radius * sin(Angle)));
+		auto ChunkX = FloorC(TowerPos.x / cChunkDef::Width);
+		if (TowerPos.x < 0)
+		{
+			ChunkX--;
+		}
+		auto ChunkZ = FloorC(TowerPos.z / cChunkDef::Width);
+		if (TowerPos.z < 0)
+		{
+			ChunkZ--;
+		}
+
+		m_TowerPos[cChunkCoords(ChunkX, ChunkZ)] = TowerPos;
+		Angle = fmod(Angle + (2.0 * M_PI / static_cast<double>(m_TowerProperties.size())), 2.0 * M_PI);
+	}
+}
+
+
+
+
+
+void cEnderDragonFightStructuresGen::GenFinish(cChunkDesc & a_ChunkDesc)
+{
+	// If not in the chunks to write
+	if ((a_ChunkDesc.GetChunkX() > 6) ||
+		(a_ChunkDesc.GetChunkX() < -6) ||
+		(a_ChunkDesc.GetChunkZ() > 6) ||
+		(a_ChunkDesc.GetChunkZ() < -6))
+	{
+		return;
+	}
+	// Places the exit portal
+	auto Coords = a_ChunkDesc.GetChunkCoords();
```

Comment:
Looks like things could be simplfiied here, if the Coords line was just moved a little bit earlier.